x/qgb/abci.go

```diff
@@ -113,3 +122,108 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
 		}
 	}
 }
+
+func PruneIfNeeded(ctx sdk.Context, k keeper.Keeper) {
+	if !k.CheckLatestAttestationNonce(ctx) ||
+		k.CheckLatestAttestationNonce(ctx) && k.GetLatestAttestationNonce(ctx) <= AttestationPruningThreshold {
+		// If the attestations nonce hasn't been initialized yet, or the nonce is not greater than the minimum
+		// number of attestations, we don't need to prune anything
+		return
+	}
+
+	lastAvailableNonce := k.GetLastPrunedAttestationNonce(ctx) + 1
+	lastAttestationInStore, found, err := k.GetAttestationByNonce(ctx, lastAvailableNonce)
+	if err != nil {
+		panic(err)
+	}
+	if !found {
+		panic(fmt.Sprintf("couldn't find attestation %s for pruning", lastAttestationInStore))
+	}
+	var lastNonceHeight uint64
+	switch lastAttestationInStore.Type() {
+	case types.DataCommitmentRequestType:
+		dc, ok := lastAttestationInStore.(*types.DataCommitment)
+		if !ok {
+			panic("couldn't cast data commitment type")
+		}
+		lastNonceHeight = dc.EndBlock
+	case types.ValsetRequestType:
+		vs, ok := lastAttestationInStore.(*types.Valset)
+		if !ok {
+			panic("couldn't cast valset type")
+		}
+		lastNonceHeight = vs.Height
+	}
+	lastUnbondingHeight := k.GetLastUnBondingBlockHeight(ctx)
+	if lastNonceHeight == lastUnbondingHeight {
+		// we don't need to do anything, we want to keep attestations up to the last unbonding height
+		return
+	}
+	if lastNonceHeight > lastUnbondingHeight {
+		// checking if it's the initial case following the startup of the chain
+		if lastNonceHeight == 1 && lastUnbondingHeight == 0 {
+			return
+		}
+		// we should never hit this case, since we will keep the attestations up to the last unbonding height
+		panic("missing attestations up to the unbonding height")
+	}
```

Comment:
We should panic as something bad happened. By default, we should always have attestations up to the last unbonding height. If the last available nonce has a height that is higher than the unbonding height, it means that we pruned attestations we shouldn't, so it's better to panic to make it easier to debug what's wrong, and also not to have a state machine behaving differently without us knowing

Reply:
tbh, its difficult to see where in the tests this gets tested for, and seems like it could be hit. my first reaction was "how is this not getting hit" :sweat_smile: .  especially since it appears that the unbonding height is not what I thought it was https://github.com/celestiaorg/celestia-app/pull/1799/files#r1199820231

You're correct that it can make sense to panic! especially if it is not in a critical pathway and lets us not return an error, or if the consequences of that thing happening are of the utmost importance (for example chain invarients in the bank module) but in this specific scenario I don't think it does. We shouldn't be risking a chain halt to ensure that we're pruning state correctly, and the qgb will likely work even if it does. and if it doesn't, then at least the chain continues until the next hardfork that fixes it.

Reply:
> tbh, its difficult to see where in the tests this gets tested for, and seems like it could be hit

It's not tested because it would require manually changing the state. If you run `abci.EndBlocker`  on every block, that case should never be hit

Reply:
> and the qgb will likely work even if it does. and if it doesn't, then at least the chain continues until the next hardfork that fixes it.

That's a good point. Maybe an error log message?

Reply:
that works for me :slightly_smiling_face:

Reply:
[0f22136](https://github.com/celestiaorg/celestia-app/pull/1799/commits/0f2213671a6725c13abc7515faa4e36eaef653cd)